File: server/src/models/ServiceEnquiry.js
import mongoose from "mongoose";

const serviceEnquirySchema = new mongoose.Schema(
  {
    customerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
      index: true,
    },
    customerName: {
      type: String,
      required: true,
      trim: true,
    },
    customerEmail: {
      type: String,
      trim: true,
      default: "",
    },
    phoneNumber: {
      type: String,
      required: false, // Optional - can be retrieved from user profile
      trim: true,
      default: "",
    },
    workType: {
      type: String,
      required: true,
      enum: ["Planing", "Resawing", "Debarking", "Sawing", "Other"],
      index: true,
    },
    // Array of log items - each item represents a different wood type/batch
    logItems: [{
      woodType: {
        type: String,
        required: true,
        trim: true,
      },
      numberOfLogs: {
        type: Number,
        required: true,
        min: 1,
      },
      thickness: {
        type: Number,
        required: true,
        min: 0,
      },
      width: {
        type: Number,
        required: true,
        min: 0,
      },
      length: {
        type: Number,
        required: true,
        min: 0,
      },
      cubicFeet: {
        type: Number,
        required: true,
        min: 0.1,
      },
    }],
    // Legacy fields for backward compatibility (deprecated, use logItems instead)
    woodType: {
      type: String,
      trim: true,
      default: "",
    },
    numberOfLogs: {
      type: Number,
      min: 1,
      default: null,
    },
    cubicFeet: {
      type: Number,
      required: true,
      min: 0.1,
    },
    // Derived processing details based on cubic feet
    processingHours: {
      type: Number, // Total processing time in hours (rounded up)
      min: 0,
      default: 0,
    },
    ratePerHour: {
      type: Number, // Billing rate per processing hour
      min: 0,
      default: 1200, // ₹1200 per hour
    },
    requestedDate: {
      type: Date,
      required: true,
      index: true,
    },
    requestedTime: {
      type: String, // Format: "HH:MM"
      required: true,
    },
    notes: {
      type: String,
      trim: true,
      default: "",
    },
    images: [{
      url: {
        type: String,
        required: true,
      },
      public_id: {
        type: String,
        default: "",
      },
    }],
    status: {
      type: String,
      enum: [
        "ENQUIRY_RECEIVED",
        "UNDER_REVIEW",
        "TIME_ACCEPTED",
        "ALTERNATE_TIME_PROPOSED",
        "SCHEDULED",
        "IN_PROGRESS",
        "COMPLETED",
        "CANCELLED",
        "REJECTED",
      ],
      default: "ENQUIRY_RECEIVED",
      index: true,
    },
    adminNotes: {
      type: String,
      trim: true,
      default: "",
    },
    // Accepted time (when admin accepts requested time)
    acceptedDate: {
      type: Date,
      default: null,
    },
    acceptedStartTime: {
      type: String,
      default: null,
    },
    acceptedEndTime: {
      type: String,
      default: null,
    },
    // Proposed alternate time (when admin proposes different time)
    proposedDate: {
      type: Date,
      default: null,
    },
    proposedStartTime: {
      type: String,
      default: null,
    },
    proposedEndTime: {
      type: String,
      default: null,
    },
    // Final scheduled time (after confirmation)
    scheduledDate: {
      type: Date,
      default: null,
    },
    scheduledTime: {
      type: String,
      default: null,
    },
    assignedScheduleId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "ServiceSchedule",
      default: null,
    },
    estimatedCost: {
      type: Number,
      min: 0,
      default: null,
    },
    actualCost: {
      type: Number,
      min: 0,
      default: null,
    },
    // Payment tracking
    paymentStatus: {
      type: String,
      enum: ["PENDING", "PAID", "FAILED"],
      default: "PENDING",
      index: true,
    },
    paymentMethod: {
      type: String,
      enum: ["NONE", "ONLINE", "OFFLINE"],
      default: "NONE",
    },
    razorpayOrderId: {
      type: String,
      trim: true,
      default: "",
    },
    razorpayPaymentId: {
      type: String,
      trim: true,
      default: "",
    },
    razorpaySignature: {
      type: String,
      trim: true,
      default: "",
    },
    offlinePaymentNote: {
      type: String,
      trim: true,
      default: "",
    },
    paymentDate: {
      type: Date,
      default: null,
    },
    completedAt: {
      type: Date,
      default: null,
    },
  },
  { timestamps: true }
);

// Compound indexes for efficient queries
serviceEnquirySchema.index({ customerId: 1, status: 1 });
serviceEnquirySchema.index({ status: 1, requestedDate: 1 });
serviceEnquirySchema.index({ createdAt: -1 });

const ServiceEnquiry = mongoose.model("ServiceEnquiry", serviceEnquirySchema);

export default ServiceEnquiry;

File: jc_timber_mobile/lib/models/service_enquiry.dart
class LogItem {
  final String woodType;
  final int numberOfLogs;
  final double thickness;
  final double width;
  final double length;
  final double cubicFeet;

  LogItem({
    required this.woodType,
    required this.numberOfLogs,
    this.thickness = 0,
    this.width = 0,
    this.length = 0,
    required this.cubicFeet,
  });

  Map<String, dynamic> toJson() => {
        'woodType': woodType,
        'numberOfLogs': numberOfLogs,
        'thickness': thickness,
        'width': width,
        'length': length,
        'cubicFeet': cubicFeet,
      };

  factory LogItem.fromJson(Map<String, dynamic> json) => LogItem(
        woodType: json['woodType'] ?? '',
        numberOfLogs: (json['numberOfLogs'] ?? 0) as int,
        thickness: (json['thickness'] ?? 0).toDouble(),
        width: (json['width'] ?? 0).toDouble(),
        length: (json['length'] ?? 0).toDouble(),
        cubicFeet: (json['cubicFeet'] ?? 0).toDouble(),
      );
}

class ServiceEnquiry {
  final String id;
  final String workType;
  final List<LogItem> logItems;
  final double cubicFeet;
  final int? processingHours;
  final String requestedDate;
  final String requestedTime;
  final String status;
  final double? estimatedCost;
  final String? customerName;
  final String? phoneNumber;
  final DateTime? createdAt;
  final String? notes;
  final String? adminNotes;
  final DateTime? acceptedDate;
  final String? acceptedStartTime;
  final String? acceptedEndTime;
  final DateTime? proposedDate;
  final String? proposedStartTime;
  final String? proposedEndTime;
  final DateTime? scheduledDate;
  final String? scheduledTime;
  final String? paymentStatus;
  final String? paymentMethod;
  final String? offlinePaymentNote;

  ServiceEnquiry({
    required this.id,
    required this.workType,
    required this.logItems,
    required this.cubicFeet,
    this.processingHours,
    required this.requestedDate,
    required this.requestedTime,
    required this.status,
    this.estimatedCost,
    this.customerName,
    this.phoneNumber,
    this.createdAt,
    this.notes,
    this.adminNotes,
    this.acceptedDate,
    this.acceptedStartTime,
    this.acceptedEndTime,
    this.proposedDate,
    this.proposedStartTime,
    this.proposedEndTime,
    this.scheduledDate,
    this.scheduledTime,
    this.paymentStatus,
    this.paymentMethod,
    this.offlinePaymentNote,
  });

  factory ServiceEnquiry.fromJson(Map<String, dynamic> json) {
    final logItemsRaw = json['logItems'] as List<dynamic>? ?? [];

     DateTime? _parseDate(dynamic v) {
      if (v == null) return null;
      try {
        return DateTime.parse(v.toString());
      } catch (_) {
        return null;
      }
    }

    return ServiceEnquiry(
      id: json['_id'] ?? json['id'] ?? '',
      workType: json['workType'] ?? '',
      logItems: logItemsRaw
          .map((e) => LogItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      cubicFeet: (json['cubicFeet'] ?? 0).toDouble(),
      processingHours: json['processingHours'] as int?,
      requestedDate: json['requestedDate'] != null
          ? (json['requestedDate'] is String
              ? json['requestedDate'] as String
              : DateTime.parse(json['requestedDate'].toString())
                  .toIso8601String()
                  .substring(0, 10))
          : '',
      requestedTime: json['requestedTime'] ?? '',
      status: json['status'] ?? '',
      estimatedCost: (json['estimatedCost'] as num?)?.toDouble(),
      customerName: json['customerName'] as String?,
      phoneNumber: json['phoneNumber'] as String?,
      createdAt: _parseDate(json['createdAt']),
      notes: json['notes'] as String?,
      adminNotes: json['adminNotes'] as String?,
      acceptedDate: _parseDate(json['acceptedDate']),
      acceptedStartTime: json['acceptedStartTime'] as String?,
      acceptedEndTime: json['acceptedEndTime'] as String?,
      proposedDate: _parseDate(json['proposedDate']),
      proposedStartTime: json['proposedStartTime'] as String?,
      proposedEndTime: json['proposedEndTime'] as String?,
      scheduledDate: _parseDate(json['scheduledDate']),
      scheduledTime: json['scheduledTime'] as String?,
      paymentStatus: json['paymentStatus'] as String?,
      paymentMethod: json['paymentMethod'] as String?,
      offlinePaymentNote: json['offlinePaymentNote'] as String?,
    );
  }
}
